src/bindings/swift/gen_swift.rs

```diff
@@ -0,0 +1,615 @@
+    }
+}
+
+// Serialization and deserialization errors.
+enum InternalError: Error {
+    // Reading the requested value would read past the end of the buffer.
+    case bufferOverflow
+    // The buffer still has data after lifting its containing value.
+    case incompleteData
+    // Unexpected tag byte for `Optional`; should be 0 or 1.
+    case unexpectedOptionalTag
+    // Unexpected integer that doesn't correspond to an enum case.
+    case unexpectedEnumCase
+}
+
+// A helper class to read values out of a byte buffer.
+class Reader {
+    let data: Data
+    var offset: Data.Index
+
+    init(data: Data) {
+        self.data = data
+        self.offset = 0
+    }
+
+    // Reads an integer at the current offset, in big-endian order, and advances
+    // the offset on success. Throws if reading the integer would move the
+    // offset past the end of the buffer.
+    func readInt<T: FixedWidthInteger>() throws -> T {
+        let range = offset..<offset + MemoryLayout<T>.size
+        guard data.count >= range.upperBound else {
+            throw InternalError.bufferOverflow
+        }
+        if T.self == UInt8.self {
+            let value = data[offset]
+            offset += 1
+            return value as! T
+        }
+        var value: T = 0
+        let _ = withUnsafeMutableBytes(of: &value, { data.copyBytes(to: $0, from: range)})
+        offset = range.upperBound
+        return value.bigEndian
+    }
+
+    // Reads a float at the current offset.
+    @inlinable
+    func readFloat() throws -> Float {
+        return Float(bitPattern: try readInt())
+    }
+
+    // Reads a float at the current offset.
+    @inlinable
+    func readDouble() throws -> Double {
+        return Double(bitPattern: try readInt())
+    }
+
+    // Indicates if the offset has reached the end of the buffer.
+    @inlinable
+    func hasRemaining() -> Bool {
+        return offset < data.count
+    }
+}
+
+// A helper class to write values into a byte buffer.
+class Writer {
+    var bytes: [UInt8]
+    var offset: Array<UInt8>.Index
+
+    init() {
+        self.bytes = []
+        self.offset = 0
+    }
+
+    // Writes an integer in big-endian order.
+    func writeInt<T: FixedWidthInteger>(_ value: T) {
+        var value = value.bigEndian
+        let _ = withUnsafeBytes(of: &value, { bytes.append(contentsOf: $0) })
+    }
+
+    @inlinable
+    func writeFloat(_ value: Float) {
+        writeInt(value.bitPattern)
+    }
+
+    @inlinable
+    func writeDouble(_ value: Double) {
+        writeInt(value.bitPattern)
+    }
+}
+
+// Types conforming to `Liftable` know how to deserialize ("lift") themselves
+// from a byte buffer. This is equivalent to the `Liftable` trait on the Rust
+// side.
+protocol Liftable {
+    static func lift(from: Reader) throws -> Self
+}
+
+// Types conforming to `Lowerable` know how to serialize ("lower") themselves
+// into a byte buffer. Equivalent to the `Lowerable` trait on the Rust side.
```

Comment:
These "lift" and "lower" names, I pinched from the Wasm Interface Types proposal, but TBH we're not really using them in a way that's very similar to how they're used there. Without wanting to rush into changing things right now, I'm curious if you found they helped or hindered understanding of how the code works? (As opposed to, say, a name like "Deserializble" which is what you've used in the comment here to explain that "Liftable" actually means).

Reply:
I found them pretty helpful! The thing that confused me at first was having `lift`, `liftFrom`, `lower`, and `lowerInto` methods when I was reading through the Kotlin generator. At first, I wondered why we needed `lift` and `lower`, and why we couldn't use a byte buffer for everything? Then I had a closer look at `ViaFfi` in the Rust support, and it clicked that `lift` and `lower` were equivalent to `try_from_ffi_value` and `into_ffi_value`, because allocating buffers when you only wanted to pass primitives would be super inefficient.

The other thing is, `Serializable` in Swift (the same as `ViaFfiUsingByteBuffer` in Rusr) feels like a bad name, since its constraint is `Liftable & Lowerable`, but `Lowerable` is for "types that know how to serialize themselves". 🤦‍♀️

Also, even though Swift really likes protocol names to be nouns and adjectives, I wonder if we should name our protocols and methods the same as Rust, for consistency's sake? Whatever we decide on, have it match in all our bindings so that it's easy to follow.